File: alligator/src/main/java/me/aartikov/alligator/commands/ForwardCommand.java
package me.aartikov.alligator.commands;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.aartikov.alligator.Screen;
import me.aartikov.alligator.animations.AnimationData;
import me.aartikov.alligator.destinations.ActivityDestination;
import me.aartikov.alligator.destinations.DialogFragmentDestination;
import me.aartikov.alligator.destinations.FragmentDestination;
import me.aartikov.alligator.exceptions.NavigationException;
import me.aartikov.alligator.navigators.ActivityNavigator;
import me.aartikov.alligator.navigators.DialogFragmentNavigator;
import me.aartikov.alligator.navigators.FragmentNavigator;


/**
 * Command implementation for {@code goForward} method of {@link me.aartikov.alligator.AndroidNavigator}.
 */
public class ForwardCommand extends BaseCommand {
	private Screen mScreen;
	@Nullable
	private AnimationData mAnimationData;

	public ForwardCommand(@NonNull Screen screen, @Nullable AnimationData animationData) {
		super(screen.getClass());
		mScreen = screen;
		mAnimationData = animationData;
	}

	@Override
	protected void executeForActivity(@NonNull ActivityDestination destination, @NonNull ActivityNavigator activityNavigator) throws NavigationException {
		activityNavigator.goForward(mScreen, destination, mAnimationData);
	}

	@Override
	protected void executeForFragment(@NonNull FragmentDestination destination, @NonNull FragmentNavigator fragmentNavigator) throws NavigationException {
		fragmentNavigator.goForward(mScreen, destination, mAnimationData);
	}

	@Override
	protected void executeForDialogFragment(@NonNull DialogFragmentDestination destination, @NonNull DialogFragmentNavigator dialogFragmentNavigator) throws NavigationException {
		dialogFragmentNavigator.goForward(mScreen, destination, mAnimationData);
	}
}

File: alligator/src/main/java/me/aartikov/alligator/commands/ForwardCommand.kt
package me.aartikov.alligator.commands

import me.aartikov.alligator.Screen
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.destinations.ActivityDestination
import me.aartikov.alligator.destinations.DialogFragmentDestination
import me.aartikov.alligator.destinations.FragmentDestination
import me.aartikov.alligator.exceptions.NavigationException
import me.aartikov.alligator.navigators.ActivityNavigator
import me.aartikov.alligator.navigators.DialogFragmentNavigator
import me.aartikov.alligator.navigators.FragmentNavigator

/**
 * Command implementation for `goForward` method of [me.aartikov.alligator.AndroidNavigator].
 */
class ForwardCommand(
    private val mScreen: Screen,
    private val mAnimationData: AnimationData?
) : BaseCommand(mScreen.javaClass) {

    @Throws(NavigationException::class)
    override fun executeForActivity(
        destination: ActivityDestination,
        activityNavigator: ActivityNavigator
    ) {
        activityNavigator.goForward(mScreen, destination, mAnimationData)
    }

    @Throws(NavigationException::class)
    override fun executeForFragment(
        destination: FragmentDestination,
        fragmentNavigator: FragmentNavigator
    ) {
        fragmentNavigator.goForward(mScreen, destination, mAnimationData)
    }

    @Throws(NavigationException::class)
    override fun executeForDialogFragment(
        destination: DialogFragmentDestination,
        dialogFragmentNavigator: DialogFragmentNavigator
    ) {
        dialogFragmentNavigator.goForward(mScreen, destination, mAnimationData)
    }
}